File: java/src/RunTests.java
import java.nio.ByteBuffer;
import java.security.MessageDigest;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Random;

/**
 * A command-line tool that runs the DropboxContentHasher tests.
 */
public class RunTests
{
    /**
     * A simple implementation, used solely to test the more complicated one.
     */
    public static byte[] referenceHasher(byte[] input, int length)
    {
        int offset = 0;
        int remaining = length;

        MessageDigest overallHasher = DropboxContentHasher.newSha256Hasher();
        MessageDigest blockHasher = DropboxContentHasher.newSha256Hasher();

        while (remaining > 0) {
            int partSize = Math.min(DropboxContentHasher.BLOCK_SIZE, remaining);
            blockHasher.update(input, offset, partSize);
            byte[] d = blockHasher.digest();
            overallHasher.update(d);

            remaining -= partSize;
            offset += partSize;
        }

        return overallHasher.digest();
    }

    public static boolean check(byte[] data, int[] chunkLengths)
    {
        System.out.println("checking " + Arrays.toString(chunkLengths));

        MessageDigest byteArrayHasher = new DropboxContentHasher();
        MessageDigest byteBufferHasher = new DropboxContentHasher();
        MessageDigest byteHasher = new DropboxContentHasher();

        ArrayList<MessageDigest> clones = new ArrayList<MessageDigest>();

        int totalLength = 0;
        for (int chunkLength : chunkLengths) {
            totalLength += chunkLength;
        }

        int offset = 0;
        for (int chunkLength : chunkLengths) {
            clones.add(clone(byteArrayHasher));

            byteArrayHasher.update(data, offset, chunkLength);
            byteBufferHasher.update(ByteBuffer.wrap(data, offset, chunkLength));
            for (int i = 0; i < chunkLength; i++) {
                byteHasher.update(data[offset+i]);
            }

            for (MessageDigest clone : clones) {
                clone.update(data, offset, chunkLength);
            }

            offset += chunkLength;
        }

        ArrayList<MessageDigest> allDigests = new ArrayList<MessageDigest>();
        allDigests.add(byteArrayHasher);
        allDigests.add(byteBufferHasher);
        allDigests.add(byteHasher);
        allDigests.addAll(clones);

        String reference = hex(referenceHasher(data, totalLength));
        boolean passed = true;

        ArrayList<String> results = new ArrayList<String>();
        for (MessageDigest digest : allDigests) {
            String result = hex(digest.digest());
            results.add(result);
            if (!result.equals(reference)) {
                passed = false;
            }
        }

        if (!passed) {
            System.out.println("- FAILED: " + reference + ", " + results);
        }

        return passed;
    }

    public static void main(String[] args)
    {
        if (args.length > 0) {
            System.err.println("No arguments expected; got " + args.length + ".");
            System.exit(1); return;
        }

        int B = DropboxContentHasher.BLOCK_SIZE;

        int[][] tests = {
            {0},
            {100},
            {100, 10},
            {B-1},
            {B},
            {B+1},

            {B-2, 1},
            {B-2, 2},
            {B-2, 3},

            {B-2, B+1},
            {B-2, B+2},
            {B-2, B+3},

            {5, 5, 5},
            {5, 5, 5, B},
            {5, 5, 5, 3*B},
            {5, 5, 5, 3*B, 5, 5, 5, 3*B},
        };

        int longestLength = 0;
        for (int[] test : tests) {
            longestLength = Math.max(longestLength, sum(test));
        }

        System.out.println("generating random data");
        byte[] data = new byte[longestLength];
        new Random(0).nextBytes(data);

        for (int[] test : tests) {
            boolean passed = check(data, test);
            if (!passed) {
                System.exit(1); return;
            }
        }

        System.out.println("all passed");
    }

    private static int sum(int[] a)
    {
        int sum = 0;
        for (int i : a) {
            sum += i;
        }
        return sum;
    }

    public static MessageDigest clone(MessageDigest v)
    {
        try {
            return (MessageDigest) v.clone();
        } catch (CloneNotSupportedException ex) {
            throw new AssertionError("Couldn't clone()", ex);
        }
    }

    static final char[] HEX_DIGITS = new char[]{
        '0', '1', '2', '3', '4', '5', '6', '7', '8', '9',
        'a', 'b', 'c', 'd', 'e', 'f'};

    public static String hex(byte[] data)
    {
        char[] buf = new char[2*data.length];
        int i = 0;
        for (byte b : data) {
            buf[i++] = HEX_DIGITS[(b & 0xf0) >>> 4];
            buf[i++] = HEX_DIGITS[b & 0x0f];
        }
        return new String(buf);
    }
}

File: csharp/src/RunTests.cs
using System;
using System.Security.Cryptography;

/// <summary>
/// A command-line tool that runs the DropboxContentHasher tests.
/// </summary>
public class RunTests
{
    public static void SubMain(String[] args)
    {
        if (args.Length > 0) {
            Console.WriteLine("The run-tests sub-command expects zero arguments; got " + args.Length + ".");
            Environment.Exit(1); return;
        }

        int B = DropboxContentHasher.BLOCK_SIZE;

        int[][] tests = {
            //new[] {0},
            //new[] {100},
            //new[] {100, 10},
            //new[] {B-1},
            //new[] {B},
            new[] {B+1},

            new[] {B-2, 1},
            new[] {B-2, 2},
            new[] {B-2, 3},

            new[] {B-2, B+1},
            new[] {B-2, B+2},
            new[] {B-2, B+3},

            new[] {5, 5, 5},
            new[] {5, 5, 5, B},
            new[] {5, 5, 5, 3*B},
            new[] {5, 5, 5, 3*B, 5, 5, 5, 3*B},
        };

        int longestLength = 0;
        foreach (var test in tests) {
            longestLength = Math.Max(longestLength, Sum(test));
        }

        Console.WriteLine("generating random data");
        byte[] data = new byte[longestLength];
        new Random(0).NextBytes(data);

        foreach (var test in tests) {
            bool passed = Check(data, test);
            if (!passed) {
                Environment.Exit(1); return;
            }
        }

        Console.WriteLine("all passed");
    }


    /// <summary>
    /// A simple implementation, used solely to test the more complicated one.
    /// </summary>
    public static byte[] ReferenceHasher(byte[] input, int length)
    {
        int offset = 0;
        int remaining = length;

        var overallHasher = SHA256.Create();
        var blockHasher = SHA256.Create();

        while (remaining > 0) {
            int partSize = Math.Min(DropboxContentHasher.BLOCK_SIZE, remaining);
            blockHasher.TransformFinalBlock((byte[])input.Clone(), offset, partSize);
            byte[] d = blockHasher.Hash;
            blockHasher.Initialize();
            overallHasher.TransformBlock(d, 0, d.Length, d, 0);

            remaining -= partSize;
            offset += partSize;
        }

        overallHasher.TransformFinalBlock(Array.Empty<byte>(), 0, 0);
        return overallHasher.Hash;
    }

    public static bool Check(byte[] data, int[] chunkLengths)
    {
        Console.WriteLine("checking [{0}]", string.Join(", ", chunkLengths));

        var hasher = new DropboxContentHasher();
        int totalLength = Sum(chunkLengths);

        int offset = 0;
        foreach (int chunkLength in chunkLengths) {
            hasher.TransformBlock(data, offset, chunkLength, data, offset);
            offset += chunkLength;
        }

        hasher.TransformFinalBlock(Array.Empty<byte>(), 0, 0);
        var result = DropboxContentHasher.ToHex(hasher.Hash);
        var reference = DropboxContentHasher.ToHex(ReferenceHasher(data, totalLength));

        bool passed = (result == reference);
        if (!passed) {
            Console.WriteLine("- FAILED: " + reference + ", " + result);
        }
        return passed;
    }

    private static int Sum(int[] a)
    {
        int sum = 0;
        foreach (int i in a) {
            sum += i;
        }
        return sum;
    }
}
